gson/src/main/java/com/google/gson/internal/bind/RecordTypeAdapterFactory.java

```diff
@@ -0,0 +1,480 @@
+            }
+            catch (IllegalAccessException | InvocationTargetException e) {
+                throw new TypeAdapterReflectionException("Failed to invoke method [getAccessor] on RecordComponent", e);
+            }
+        }
+        <A extends Annotation> A getAnnotation(Object recordComponent, Class<A> annotation) {
+            try {
+                return (A) getAnnotation.invoke(recordComponent, annotation);
+            }
+            catch (IllegalAccessException | InvocationTargetException e) {
+                throw new TypeAdapterReflectionException(
+                        "Failed to invoke method [getAnnotation] on RecordComponent", e);
+            }
+        }
+
+
+        public <T> Field getField(Object recordComponent, TypeToken<T> type) {
+            // There is nothing in the RecordComponent class to access the underlying field, so we do this based
+            // on the component name.
+            String fieldName = getName(recordComponent);
+            try {
+                return type.getRawType().getDeclaredField(fieldName);
+            }
+            catch (NoSuchFieldException e) {
+                throw new TypeAdapterReflectionException(
+                        "Expected to find field [" + fieldName + "] on recordComponent [" + recordComponent + "]"
+                        + " on type [" + type + "]. Somehow there is a discrepancy between the record components"
+                        + " and the fields on the Class.", e);
+            }
+        }
+
+    }
+
+    @Override
+    public <T> TypeAdapter<T> create(Gson gson, TypeToken<T> type) {
+        // ?: Is this not a record?
+        if (!RECORD_HELPER.isRecord(type)) {
+            // Yes -> Not supported, we return null as per the TypeAdapterFactory contract
+            return null;
+        }
+
+        try {
+            // To construct record fields, we actually need to do some reflection on the Gson instance. This
+            // logic is contained in the RecordFieldFactory.
+            RecordFieldFactory recordFieldFactory =
+                    new RecordFieldFactory(gson, excluder, jsonAdapterFactory, constructorConstructor);
+
+            Object[] recordComponents = RECORD_HELPER.getRecordComponents(type);
+            Class[] componentTypes = new Class[recordComponents.length];
+            for (int i = 0; i < recordComponents.length; i++) {
+                componentTypes[i] = RECORD_HELPER.getType(recordComponents[i]);
+            }
+
+            // Find the canonical constructor on the Record that corresponds to the record components.
+            // There is no method in the Java API to do the equivialent, instead we rely on the constructor and
+            // recordComponent order to be the same. This construct matches the StackOverflow answer here:
+            // https://stackoverflow.com/a/67127067
+            Constructor<T> recordConstructor = (Constructor<T>) type.getRawType().getConstructor(componentTypes);
```

Comment:
The implicit canonical constructor has the same visibility as the record class, see https://docs.oracle.com/javase/specs/jls/se17/html/jls-8.html#jls-8.10.4.
So maybe it would make sense to try make the constructor accessible to support non-public records as well without requiring the user to specify an explicit `public` constructor. Possibly in a similar way to how [`ConstructorConstructor.newDefaultConstructor`](https://github.com/google/gson/blob/2154e468b490c8a6576db13eca31c9d1afe6bf78/gson/src/main/java/com/google/gson/internal/ConstructorConstructor.java#L185) does it (with the `ReflectionAccessFilter` checks).